sh-runtime.c

```diff
@@ -673,115 +544,97 @@ DEPENDS_ON(putchar)
   putstr("    : $((__addr += 1))\n");
   putstr("  done\n");
   putstr("}\n");
-END_RUNTIME_FUN(print_pnut_str)
+END_RUNTIME_FUN(put_pstr)
 
-DEFINE_RUNTIME_FUN(print_string)
-DEPENDS_ON(int_to_char)
-  putstr("# Emit a C-string line by line so that whitespace isn't mangled\n");
-  putstr("print_string() {\n");
-  putstr("  __addr=$1; shift\n");
-  putstr("  __max_len=100000000\n");
-  putstr("  __delim=0\n");
-  putstr("  __len=0\n");
-  putstr("  __acc=\"\"\n");
-  putstr("  if [ $# -ge 1 ] ; then __delim=$1   ; shift ; fi # Optional end of string delimiter\n");
-  putstr("  if [ $# -ge 1 ] ; then __max_len=$1 ; shift ; fi # Optional max length\n");
-  putstr("  while [ $((_$__addr)) != $__delim ] && [ $__max_len -gt $__len ] ; do\n");
-  putstr("    __char=$((_$__addr))\n");
-  putstr("    __addr=$((__addr + 1))\n");
-  putstr("    __len=$((__len + 1))\n");
-  putstr("    case $__char in\n");
-  putstr("      10) # 10 == '\\n'\n");
-  putstr("        printf \"%s\\n\" \"$__acc\"\n");
-  putstr("        __acc=\"\" ;;\n");
-  putstr("      *)\n");
-  call_int_to_char("        ", "$__char")
-  putstr("        __acc=\"$__acc$__char\" ;;\n");
-  putstr("    esac\n");
+DEFINE_RUNTIME_FUN(printf)
+DEPENDS_ON(put_pstr)
+  putstr("read_int() {\n");
+  putstr("  __int=\n");
+  putstr("  while [ $((_$__fmt_ptr)) != 0 ] && [ $((_$__fmt_ptr)) -ge 48 ] && [ $((_$__fmt_ptr)) -le 57 ]; do\n");
+  putstr("    __int=\"$__int$((_$__fmt_ptr - 48))\"\n");
+  putstr("    : $((__fmt_ptr += 1))\n");
   putstr("  done\n");
-  putstr("  printf \"%s\" \"$__acc\"\n");
   putstr("}\n");
-END_RUNTIME_FUN(print_string)
-
-DEFINE_RUNTIME_FUN(printf)
-DEPENDS_ON(print_string)
-DEPENDS_ON(pack_string)
-DEPENDS_ON(int_to_char)
+  putstr("\n");
+  putstr("invalid_format_error() {\n");
+  putstr("  printf \"Invalid format specifier. %%\"\n");
+  putstr("  : $((_$__fmt_ptr = 0)) # Terminate string after %...\n");
+  putstr("  _put_pstr __ $__mod_start\n");
+  putstr("  printf \"\\n\"\n");
+  putstr("  exit 1\n");
+  putstr("}\n");
+  putstr("\n");
   putstr("_printf() { # $1 = printf format string, $2... = printf args\n");
   putstr("  : $(($1 = 0)); shift # Return 0\n");
   putstr("  __fmt_ptr=$1; shift\n");
+  putstr("  __mod_start=0\n");
   putstr("  __mod=0\n");
+  putstr("  __flags=\n");
+  putstr("  __width=\n");
+  putstr("  __width=\n");
+  putstr("  __precision=\n");
```

Comment:
Instead of trying to parse everything at once, this new `printf` implementation processes each part of the format specifier in their own iteration. Each iteration advances the state machine encoded by those variables until it hits a terminal case (`'d'|'i'|'o'|'u'|'x'|'X'`) in which case the shell's `printf` function is called and the machine resets.